File: Problems/Algorithms/72. Edit Distance/EditDistance.kt
class Solution {
    fun minDistance(word1: String, word2: String): Int {
        val n1 = word1.length
        val n2 = word2.length
        val dp = Array(n1+1) { IntArray(n2+1) { 0 } }
        
        for (i in 1..n1) {
            dp[i][0] = i
        }
        for (j in 1..n2) {
            dp[0][j] = j
        }
        
        for (i in 1..n1) {
            for (j in 1..n2) {
                if (word1.get(i-1) == word2.get(j-1)) {
                    dp[i][j] = dp[i-1][j-1]
                } else {
                    dp[i][j] = 1 + minOf(minOf(dp[i-1][j], dp[i][j-1]), dp[i-1][j-1])
                }
            }
        }
        
        return dp[n1][n2]
    }
}

File: Problems/Algorithms/72. Edit Distance/edit_distance.rs
impl Solution {
    pub fn min_distance(word1: String, word2: String) -> i32 {
        let (s1, s2) = (word1.as_bytes(), word2.as_bytes());
        let (n1, n2) = (word1.len(), word2.len());
        let mut dp = vec![vec![0; n2+1]; n1+1];
        
        for i in 1..=n1 {
            dp[i][0] = i;
        }
        for j in 1..=n2 {
            dp[0][j] = j;
        }
        
        for i in 1..=n1 {
            for j in 1..=n2 {
                if s1[i-1] == s2[j-1] {
                    dp[i][j] = dp[i-1][j-1];
                } else {
                    dp[i][j] = 1 + dp[i-1][j-1].min(dp[i-1][j].min(dp[i][j-1]));
                }
            }
        }
        
        dp[n1][n2] as i32
    }
}

